### PR TITLE
feat(jsonrpc): parse and strip the ln prefix in the json parse 

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Currently, we have a check in place for the presence of the `lightning:` 
while decoding a bolt11 string. 
However, this is causing some issues with
the user experience. Therefore, we should strip this prefix
before calling each core command of core lightning.

our code is tight with this strip when it is needed so,
my approach in the first commit is to fix just the problem,
and then fix the source of the problem by parsing the string
without `lightning:` prefix inside the argument parsing.

The reason for this is that we also store this prefix
 in the database, so we need to deal with it accordingly.

See commits for more details

fixes https://github.com/ElementsProject/lightning/issues/6207

### CODE REVIEW

**Comment by vincenzopalazzo** [comment]
Ok we need to lowering the old invoice string inside the db too, as @rustyrussell suggested on discord

File: ccan/ccan/tal/str/str.h [hunk 1]
```diff
@@ -222,4 +222,14 @@ char *tal_strjoin_(const void *ctx,
...
 		 const char *label, const char *regex, ...);
+
+/**
+ * tal_strlowering - match/extract from a string via (extended) regular expressions.
+ * @ctx: the context to tal from (often NULL)
+ * @string: the string to try to match (can be take())
+ */
+#define tal_strlowering(ctx, string)					\
+	tal_strlowering_(ctx, string)
+char *tal_strlowering_(const void *ctx, const char *string TAKES);
```
Comment (rustyrussell) [inline ccan/ccan/tal/str/str.h:233]: This does not need to be macro-wrapped, and can simply go in common/utils.c for now (perhaps with a comment that it belongs in ccan).  We can promote it separately.

File: common/bolt11.c [hunk 2]
```diff
@@ -678,6 +679,28 @@ static bool bech32_decode_alloc(const tal_t *ctx,
...
+const char *to_canonical_invstr(const tal_t *ctx,
+				const char *invstring)
+{
+	/* BOLT #11:
+	 *
+	 * If a URI scheme is desired, the current recommendation is to either
+	 * use 'lightning:' as a prefix before the BOLT-11 encoding */
+	if (strstarts(invstring, "lightning:") || strstarts(invstring, "LIGHTNING:"))
+		invstring += strlen("lightning:");;
+	return tal_strlowering(ctx, invstring);
```
Comment (rustyrussell) [inline common/bolt11.c:691]: if (has_lightning_prefix(invstring)?

File: common/bolt11.c [hunk 3]
```diff
@@ -678,6 +679,28 @@ static bool bech32_decode_alloc(const tal_t *ctx,
...
+}
+
+inline static bool is_canonical_invstr(const char *invstring)
```
Comment (rustyrussell) [inline common/bolt11.c:694]: No need for `inline` here, that's the compilers' decision.

File: common/bolt11.c [hunk 4]
```diff
@@ -678,6 +679,28 @@ static bool bech32_decode_alloc(const tal_t *ctx,
...
+{
+	/* BOLT #11:
+	 *
+	 * If a URI scheme is desired, the current recommendation is to either
+	 * use 'lightning:' as a prefix before the BOLT-11 encoding */
+	return !(strstarts(invstring, "lightning:") ||
+		strstarts(invstring, "LIGHTNING:"));
+}
```
Comment (rustyrussell) [inline common/bolt11.c:702]: This should be inverted and called "has_lightning_prefix()"?

File: common/bolt11.h [hunk 5]
```diff
@@ -104,6 +104,7 @@ struct bolt11 *bolt11_decode_nosig(const tal_t *ctx, const char *str,
 				   bool *have_n,
 				   char **fail);
 
+
 /* Initialize an empty bolt11 struct with optional amount */
 struct bolt11 *new_bolt11(const tal_t *ctx,
```
Comment (rustyrussell) [inline common/bolt11.h:109]: Unnecessary whitespace change.

File: common/json_param.c [hunk 6]
```diff
@@ -439,6 +439,20 @@ struct command_result *param_string(struct command *cmd, const char *name,
...
 
+struct command_result *param_invstring(struct command *cmd, const char *name,
+				       const char * buffer, const jsmntok_t *tok,
+				       const char **str)
+{
+	/* FIXME: there are a couple of things here
+	 * - we should really have in ccan a function to strip a prefix;
+	 * - I am fall in some check bunds error with tal so I can not
+	 * reuse the same *str */
+	const char *strtmp = tal_strndup(cmd, buffer + tok->start,
```
Comment (rustyrussell) [inline common/json_param.c:450]: json_strdup() is neater, and this leaks.  Allocate off NULL, then use take() below:

File: common/json_param.c [hunk 7]
```diff
@@ -439,6 +439,20 @@ struct command_result *param_string(struct command *cmd, const char *name,
...
+			   tok->end - tok->start);
+	*str = to_canonical_invstr(cmd, strtmp);
```
Comment (rustyrussell) [inline common/json_param.c:452]: *str = to_canonical_invstr(cmd, take(strtmp))

File: wallet/db.c [hunk 8]
```diff
@@ -1688,3 +1694,75 @@ static void migrate_invalid_last_tx_psbts(struct lightningd *ld,
...
+		struct db_stmt *update_stmt;
+		const char *invstr;
+		u64 id;
+
+		id = db_col_u64(stmt, "id");
+		invstr = db_col_arr(tmpctx, stmt, "bolt11", char);
+		/* For longer explanation see migration for
+		 * invoice table
+		 *
+		 * In short, `bolt11` can be null with older nodes.*/
```
Comment (rustyrussell) [inline wallet/db.c:1753]: You can use sendpay/sendonion directly I think, and thus not set bolt11?

**Review by rustyrussell — CHANGES_REQUESTED** [review]
You also missed some places in invoice.c where the parameter is called "invstring".

Mainly, this should be split into three commitments:

1. Infrastructure.  Add canonicalization and parsing functions.
2. Changing all the parameters, adding test.  Add a Changelog-Changed: line here!
3. The database migration.

The concept, however, is good!

File: common/bolt11.c (hunk 4 above)
Comment (vincenzopalazzo) [reply common/bolt11.c:702]: Much better yeah this logic was a little bit brain stressing :)

File: common/json_param.c (hunk 6 above)
Comment (vincenzopalazzo) [reply common/json_param.c:450]: Ah right, this is the bounds that I was breaking, mh I guess this is something that I did not understand of ccan, I need to scan again the code :)

Thanks for the mem check fix :)

File: wallet/db.c (hunk 8 above)
Comment (vincenzopalazzo) [reply wallet/db.c:1753]: Ah right, yeah, I need to update my storytelling in the comments. Thanks!

I always forget how many method core lightning has to do a payment!

**Comment by vincenzopalazzo** [comment]
Ok making some house keeping and reorganising the commit as @rustyrussell suggested.

**Comment by rustyrussell** [comment]
This is looking good, but needs tests!

**Comment by vincenzopalazzo** [comment]
>This is looking good, but needs tests!

Mh, the `test_pay.py` file contains a test, and then we test the migration db inside the db_migration as well?

Maybe I would add some more stress test from the db migration?

**Comment by vincenzopalazzo** [comment]
Ok looks like there is a Ci failure regarding the sanitiser

```
----------------------------- Captured stderr call -----------------------------
=================================================================
==77454==ERROR: AddressSanitizer: heap-use-after-free on address 0x6130000ba9d8 at pc 0x5573b4bd04a1 bp 0x7ffd64c681c0 sp 0x7ffd64c681b8
READ of size 8 at 0x6130000ba9d8 thread T0
    #0 0x5573b4bd04a0 in to_tal_hdr /home/runner/work/lightning/lightning/ccan/ccan/tal/tal.c:178:40
    #1 0x5573b4bd007c in tal_free /home/runner/work/lightning/lightning/ccan/ccan/tal/tal.c:514:17
    #2 0x5573b48c19bb in str_lowering /home/runner/work/lightning/lightning/common/utils.c:243:2
    #3 0x5573b480d6f1 in to_canonical_invstr /home/runner/work/lightning/lightning/common/bolt11.c:697:9
    #4 0x5573b486ab8e in param_invstring /home/runner/work/lightning/lightning/common/json_param.c:449:9
    #5 0x5573b4875fff in make_callback /home/runner/work/lightning/lightning/common/json_param.c:64:9
    #6 0x5573b487503c in parse_by_name /home/runner/work/lightning/lightning/common/json_param.c:169:10
    #7 0x5573b4868a4a in param_arr /home/runner/work/lightning/lightning/common/json_param.c:290:10
    #8 0x5573b48695ec in param /home/runner/work/lightning/lightning/common/json_param.c:375:9
    #9 0x5573b45c2fe0 in json_decodepay /home/runner/work/lightning/lightning/lightningd/invoice.c:1531:7
    #10 0x5573b45d993e in command_exec /home/runner/work/lightning/lightning/lightningd/jsonrpc.c:657:8
    #11 0x5573b45cedbe in rpc_command_hook_final /home/runner/work/lightning/lightning/lightningd/jsonrpc.c:785:21
    #12 0x5573b4738646 in plugin_hook_call_ /home/runner/work/lightning/lightning/lightningd/plugin_hook.c:288:3
    #13 0x5573b45e24ab in plugin_hook_call_rpc_command /home/runner/work/lightning/lightning/lightningd/jsonrpc.c:873:1
    #14 0x5573b45e18eb in parse_request /home/runner/work/lightning/lightning/lightningd/jsonrpc.c:981:14
    #15 0x5573b45dc6e8 in read_json /home/runner/work/lightning/lightning/lightningd/jsonrpc.c:1086:2
    #16 0x5573b4b72c85 in next_plan /home/runner/work/lightning/lightning/ccan/ccan/io/io.c:59:9
    #17 0x5573b4b77344 in do_plan /home/runner/work/lightning/lightning/ccan/ccan/io/io.c:407:10
    #18 0x5573b4b76933 in io_ready /home/runner/work/lightning/lightning/ccan/ccan/io/io.c:417:8
    #19 0x5573b4b82e1e in io_loop /home/runner/work/lightning/lightning/ccan/ccan/io/poll.c:453:5
    #20 0x5573b45ca679 in io_loop_with_timers /home/runner/work/lightning/lightning/lightningd/io_loop_with_timers.c:22:12
    #21 0x5573b45ea2bd in main /home/runner/work/lightning/lightning/lightningd/lightningd.c:1242:23
    #22 0x7f5ec19d9d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #23 0x7f5ec19d9e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #24 0x5573b43fbbe4 in _start (/home/runner/work/lightning/lightning/lightningd/lightningd+0x3adbe4) (BuildId: 05bcd51a80b42df115b92f1d944120e5343643a2)
```

File: wallet/db.c [hunk 9]
```diff
@@ -1689,3 +1695,73 @@ static void migrate_invalid_last_tx_psbts(struct lightningd *ld,
...
+	stmt = db_prepare_v2(db, SQL("SELECT bolt11, id"
+				     " FROM invoices;"));
+	db_query_prepared(stmt);
+	while (db_step(stmt)) {
+		struct db_stmt *update_stmt;
+		const char *invstr;
+		u64 id;
+
+		id = db_col_u64(stmt, "id");
+		invstr = db_col_arr(tmpctx, stmt, "bolt11", char);
```
Comment (rustyrussell) [inline wallet/db.c:1718]: db_col_strdup().

File: wallet/db.c [hunk 10]
```diff
@@ -1689,3 +1695,73 @@ static void migrate_invalid_last_tx_psbts(struct lightningd *ld,
...
+		/* `bolt11` can be null in some cases like:
+		 * - The older payments has the bolt11 missing;
+		 * - The payment is done without bolt11, such as
+		 * using sendpay/sendonion` directly.*/
+		if (invstr) {
```
Comment (rustyrussell) [inline wallet/db.c:1753]: Use "WHERE bolt11 IS NOT null" in select instead?

File: wallet/db.c [hunk 11]
```diff
@@ -1689,3 +1695,73 @@ static void migrate_invalid_last_tx_psbts(struct lightningd *ld,
...
+	stmt = db_prepare_v2(db, SQL("SELECT bolt11, id"
+				     " FROM payments;"));
+	db_query_prepared(stmt);
+	while (db_step(stmt)) {
+		struct db_stmt *update_stmt;
+		const char *invstr;
+		u64 id;
+
+		id = db_col_u64(stmt, "id");
+		invstr = db_col_arr(tmpctx, stmt, "bolt11", char);
```
Comment (rustyrussell) [inline wallet/db.c:1748]: db_col_strdup()?

File: common/utils.c [hunk 12]
```diff
@@ -239,7 +239,6 @@ char *str_lowering(const void *ctx, const char *string TAKES)
 	char *ret;
 
 	ret = tal_strdup(ctx, string);
-	for(char *p = ret; *p; p++) *p=tolower(*p);
-	tal_free(string);
+	for (char *p = ret; *p; p++) *p=tolower(*p);
```
Comment (rustyrussell) [inline common/utils.c:242]: Ah, good, please fold into previous commit!

**Review by rustyrussell — COMMENTED** [review]
Minor changes only, looks good!

File: common/utils.h [hunk 13]
```diff
@@ -160,4 +160,13 @@ extern const tal_t *wally_tal_ctx;
  * Returns created temporary path name at *created if successful. */
 int tmpdir_mkstemp(const tal_t *ctx, const char *template TAKES, char **created);
 
+/**
+ * tal_strlowering - match/extract from a string via (extended) regular expressions.
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply common/utils.h:164]: copy and paste :) sorry

File: common/utils.c (hunk 12 above)
Comment (vincenzopalazzo) [reply common/utils.c:242]: yeah it was in a fixup commit just testing the code in the CI

File: wallet/db.c (hunk 10 above)
Comment (vincenzopalazzo) [reply wallet/db.c:1753]: Yeah much better

**Comment by vincenzopalazzo** [comment]
Ok rebased an applied the @rustyrussell suggestion.

Now just see if the CI is happy!

**Review by rustyrussell — APPROVED** [review]

**Pull request merged**